modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ConcatPublisher.java

```diff
@@ -0,0 +1,263 @@
+import java.lang.invoke.VarHandle;
+import java.util.Iterator;
+import java.util.concurrent.Flow.Publisher;
+import java.util.concurrent.Flow.Subscriber;
+import java.util.concurrent.Flow.Subscription;
+import java.util.concurrent.atomic.AtomicInteger;
+
+
+/**
+ * Thread-safe realization of combine multiple publishers. Generally, start consuming a source once the previous source has terminated,
+ * building a chain.
+ */
+public class ConcatPublisher<T> implements Publisher<T> {
+    /** Iterator of upstream publishers. */
+    private final Iterator<Publisher<? extends T>> sources;
+
+    /**
+     * Constructor.
+     *
+     * @param sources Iterator of upstream publishers.
+     */
+    public ConcatPublisher(Iterator<Publisher<? extends T>> sources) {
+        this.sources = sources;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void subscribe(Subscriber<? super T> downstream) {
+        ConcatCoordinator<T> subscription = new ConcatCoordinator<>(downstream, sources);
+
+        downstream.onSubscribe(subscription);
+        subscription.drain();
+    }
+
+    /**
+     * Concatenation composite subscription.
+     */
+    static final class ConcatCoordinator<T> extends SubscriptionArbiter implements Subscriber<T> {
+        /** Counter to prevent concurrent execution of a critical section. */
+        private final AtomicInteger guardCntr = new AtomicInteger();
+
+        final Subscriber downstream;
+
+        final Iterator<Publisher<? extends T>> sources;
+
+        long consumed;
+
+        ConcatCoordinator(
+                Subscriber downstream,
+                Iterator<Publisher<? extends T>> sources
+        ) {
+            this.downstream = downstream;
+            this.sources = sources;
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            setSubscription(s);
+        }
+
+        @Override
+        public void onNext(T item) {
+            consumed++;
+            downstream.onNext(item);
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            downstream.onError(throwable);
+        }
+
+        @Override
+        public void onComplete() {
+            drain();
+        }
+
+        void drain() {
+            if (guardCntr.getAndIncrement() == 0) {
+                do {
+                    if (isCancelled()) {
+                        return;
+                    }
+
+                    if (!sources.hasNext()) {
+                        downstream.onComplete();
+                        return;
+                    }
+
+                    long c = consumed;
+                    if (c != 0L) {
+                        consumed = 0L;
+                        setProduced(c);
+                    }
+
+                    sources.next().subscribe(this);
+
+                } while ((guardCntr.getAndDecrement() - 1) != 0);
+            }
+        }
+    }
+
+    static class SubscriptionArbiter implements Subscription {
```

Comment:
We currently only have one class that inherits from `SubscriptionArbiter`.
It seems we should just combine `ConcatCoordinator` with `SubscriptionArbiter`